### PR TITLE
Tagging

### PR DESCRIPTION
**Pull request by langjoh**
Multiple tag options from check box possible now. UX/UI can be improved in the future

### CODE REVIEW

**Comment by aydamacink** [comment]
We need to stub The API request tests and also write some tests for the tagging feature

File: spec/factories/taggings.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
+  factory :tagging do
+    drop nil
+tag nil
```
Comment (berlintam) [inline spec/factories/taggings.rb]: can u fix the indentation here pls? thanks!

File: spec/features/stats_page_spec.rb [hunk 2]
```diff
@@ -1,11 +0,0 @@
-require 'spec_helper'
-require 'rails_helper'
-
-describe "Stats page", type: :feature, :vcr => {:cassette_name => "place" } do
-
-  it "should have a highchart", js: true do
-    visit "/stats"
-    expect(page).to have_css(".highcharts-container")
-  end
-
```
Comment (berlintam) [inline spec/features/stats_page_spec.rb:10]: hmmmm I think we were going to keep the stats page (bc it's interesting & could be used later at some point). Can you leave this spec in here? Or was it somehow messing stuff up?

**Comment by berlintam** [comment]
This is awesome & it totally works! Thanks so much! I love the checkboxes! eeeeeeee!

A few things that I think need to be done before it can be merged:
1. tests (which you already mentioned above)
2. can you pls list the tags related to the drop on the drop#show page? It doesn't have to be anything fancy, just something like Tags: sound, tour. Perhaps below the coordinates (underneath the track/map)
3. can you make it possible for a user to edit the tags for their drop? If you own a drop you're able to now edit the drop's place information, and I think it should also be possible to edit the tags. To do this, I think you'll have to add the tagging stuff to the `edit.html.erb` and also probably the update method in the controller. 

If point 3 proves to be too difficult/annoying, you can skip it. I know that it's important to each of you that you complete your individual features before the summer is up, so if that's the case just let me know & we can create a new issue for it. 

excited

File: app/views/drops/_form.html.erb [hunk 3]
```diff
@@ -36,9 +36,27 @@
...
+
+    <div class='row'>
+      <div class='row'>
+        <div class='title'>
+          <h2>Third, tag your drop:</h2>
+        </div>
+      </div>
+      <div class='row'>
+      <%= label_tag(:tag, "Please select one or more tags") %></br>
+      <% Tag.all.each do |tag| -%>
```
Comment (berlintam) [inline app/views/drops/_form.html.erb]: what's this minus sign here? `-%>`

Comment (aydamacink) [reply app/views/drops/_form.html.erb]: a mistake , thanks 4 catching it, now it is fixed :)

File: spec/factories/taggings.rb (hunk 1 above)
Comment (aydamacink) [reply spec/factories/taggings.rb]: done

File: app/controllers/drops_controller.rb [hunk 4]
```diff
@@ -69,8 +69,8 @@ def upvote
   private
 
     def set_drop
-      @drop = Drop.find(params[:id].to_i)
-    end
+      page_not_found unless @drop = Drop.find_by(id: params[:id].to_i)
```
Comment (berlintam) [inline app/controllers/drops_controller.rb:72]: I'm not sure if you can do this. I think instead it should be:

`@drop = Drop.find_by(id: params[:id].to_i) unless page_not_found` 

that being said I'm not totally 100% sure where `page_not_found` comes from.... but give that a shot, & if it doesn't work, try writing this a different way. You can play around w/ binding.pry too & see what happens here.

Comment (aydamacink) [reply app/controllers/drops_controller.rb:72]: cool Thanks for your comment, we figured it out it was `(:find_by).with({ id: drop.id })`

But then through a suggestion from @carpodaster we refactored it because the test shouldn't care about the methods by which the controller extracts data from the database

Comment (carpodaster) [reply app/controllers/drops_controller.rb:72]: The `page_not_found` shouldn't be necessary: Rails' application controller rescues from `ActiveRecord::RecordNotFound` exceptions by displaying the 404 error page  and returning the 404 status code _in production_. In dev mode, it will still show the error.

The error is raised when you use `AR::Base.find(id)` or `find_by!(id: id)` with an _id_ that doesn't exist.

``` ruby
def set_drop
  @drop = Drop.find(params[:id])
end
```

Comment (aydamacink) [reply app/controllers/drops_controller.rb:72]: @carpodaster: Thanks for this explanation. It was really great, but if we want to show the error with a pretty nice form with cats and all that, we would need this method right?

File: spec/features/stats_page_spec.rb (hunk 2 above)
Comment (aydamacink) [reply spec/features/stats_page_spec.rb:10]: We  brought it back :)

**Comment by langjoh** [comment]
@berlintam We just completed the tests and reintroduced the highcharts spec, so the PR might be ready to merge? ;) Or should we wait until the tag search is ready to merge, as it is a sub-branch of tagging?

**Comment by berlintam** [comment]
awesome job!!! 🎆 
This can totally be merged prior to the tag-search branch, in fact, it'll have to be merged before tag-search (otherwise things'll get messed up). 
I can take a look at it this evening/tomorrow if that's cool?

**Comment by aydamacink** [comment]
We left the edit tags function for a later time, so @Powerglove will make an issue out of it :)

File: app/models/tag.rb [hunk 5]
```diff
@@ -0,0 +1,4 @@
+class Tag < ActiveRecord::Base
+    has_many :taggings
+    has_many :drops, through: :taggings
```
Comment (berlintam) [inline app/models/tag.rb]: can you fix the indentation here pls?

File: db/migrate/20160915144154_add_color_to_tags.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddColorToTags < ActiveRecord::Migration
+  def change
+    add_column :tags, :color, :string, default: "#f50"
```
Comment (berlintam) [inline db/migrate/20160915144154_add_color_to_tags.rb:3]: what's the benefit of having the colour inside the model itself as an attribute? Why not just style it w/ css?

File: spec/controllers/drops_controller_spec.rb [hunk 7]
```diff
@@ -99,12 +98,11 @@
...
 
     context 'unsuccessfully updating a drop' do
-      let(:drop)   { build_stubbed :drop, :with_place, id: 2, sc_user_id: 654 }
-      let(:params) {{ id: 2, "place" => { "name" => "a sneaky hacker place", "location" => "sneaky hacker location" }}}
+      let(:drop)   { create :drop, :with_place, sc_user_id: 654 }
+      let(:params) {{ id: drop.id, "place" => { "name" => "a sneaky hacker place", "location" => "sneaky hacker location" }}}
 
       before do
         expect(controller).to receive(:current_user).and_return nil
-        expect(Drop).to receive(:find).with(drop.id).and_return drop
```
Comment (berlintam) [inline spec/controllers/drops_controller_spec.rb:107]: why did you remove this?

File: spec/rails_helper.rb [hunk 8]
```diff
@@ -55,3 +55,31 @@
   config.include FactoryGirl::Syntax::Methods
 
 end
+
+Geocoder.configure(lookup: :test)
```
Comment (berlintam) [inline spec/rails_helper.rb]: Can you add a small comment here about why you added this? (it's to fix some geocoding issue, right?)

File: app/views/drops/_form.html.erb [hunk 9]
```diff
@@ -36,9 +36,27 @@
...
       </div>
+
+    <div class='row'>
+      <div class='row'>
+        <div class='title'>
+          <h2>Third, tag your drop:</h2>
+        </div>
+      </div>
+      <div class='row'>
+      <%= label_tag(:tag, "Please select one or more tags") %></br>
```
Comment (berlintam) [inline app/views/drops/_form.html.erb]: Do we require users to select tags? Should we?

File: spec/rails_helper.rb (hunk 8 above)
Comment (carpodaster) [reply spec/rails_helper.rb]: It almost already resolves #106 when merged

File: app/views/drops/_form.html.erb (hunk 9 above)
Comment (langjoh) [reply app/views/drops/_form.html.erb]: you're right, we could leave it up to them.

File: spec/controllers/drops_controller_spec.rb (hunk 7 above)
Comment (aydamacink) [reply spec/controllers/drops_controller_spec.rb:107]: In  app/controllers/drops_controller.rb 

we changed the #set_drop method and are no longer using the find method we were using there before, because we are now using  `page_not_found unless @drop = Drop.find_by(id: params[:id].to_i)`.  We didn't include this new `find_by` method in the test, because Carsten explained that the test shouldn't care about the methods by which the controller extracts data from the database, so the tests should test the outer layer more like black-box testing.

File: spec/rails_helper.rb (hunk 8 above)
Comment (aydamacink) [reply spec/rails_helper.rb]: Our build was failing because we hadn't stubbed the calls to the Google API and it was making real requests and it reached a limit, not allowing us to make more requests. So now that it is stubbed, it doesn't make real requests and you don't need to be online to run the test.

File: app/models/tag.rb (hunk 5 above)
Comment (langjoh) [reply app/models/tag.rb]: Done :)

File: app/views/drops/_form.html.erb (hunk 9 above)
Comment (aydamacink) [reply app/views/drops/_form.html.erb]: changed :) ,@tam: pls check the wording

File: db/migrate/20160915144154_add_color_to_tags.rb (hunk 6 above)
Comment (aydamacink) [reply db/migrate/20160915144154_add_color_to_tags.rb:3]: So we could also have done it with CSS, but in this way it is default and it is already there, so we don't have to check if it is defined, because if we forget it, then the tags would have no background color. It is a default that makes sure we don't have to define it and that there is a background color per se.

File: app/views/drops/_form.html.erb [hunk 10]
```diff
@@ -44,7 +44,7 @@
         </div>
       </div>
       <div class='row'>
-      <%= label_tag(:tag, "Please select one or more tags") %></br>
+      <%= label_tag(:tag, "If you want you can select one or more tags for your drop.") %></br>
```
Comment (berlintam) [inline app/views/drops/_form.html.erb:47]: 👍

File: spec/rails_helper.rb [hunk 11]
```diff
@@ -56,7 +56,7 @@
 
 end
 
-Geocoder.configure(lookup: :test)
+Geocoder.configure(lookup: :test) #stub data to prevent tests form making actual requests
```
Comment (berlintam) [inline spec/rails_helper.rb:59]: 👍

File: db/migrate/20160915144154_add_color_to_tags.rb (hunk 6 above)
Comment (berlintam) [reply db/migrate/20160915144154_add_color_to_tags.rb:3]: hmmmmmm I'm super on the fence about this..... but.... I'm willing to give it a shot..... 😉

**Comment by berlintam** [comment]
bravo
